api.go

```diff
@@ -659,9 +659,7 @@ func (c Client) newRequest(method string, metadata requestMetadata) (req *http.R
 		// Set sha256 sum for signature calculation only with signature version '4'.
 		shaHeader := unsignedPayload
 		if !c.secure {
-			if metadata.contentSHA256Bytes == nil {
-				shaHeader = hex.EncodeToString(sum256([]byte{}))
-			} else {
+			if len(metadata.contentSHA256Bytes) > 0 {
```

Comment:
But we are breaking backward compatibility here. The example is wrong it is expected in signature-v4 for PUT you are passing in the payload checksum. So setting it as `nil` is wrong. 

What should be set is perhaps `[]byte("UNSIGNED-PAYLOAD")`  is passed in from PutObject() perspective indicating that the payload checksum is going to be unsigned and internal code differentiates this. 

Otherwise we are breaking backward compatibility i.e for some servers which do not implement UNSIGNED-PAYLOAD verification for GET, HEAD operations would lead to signature issues.